### PR TITLE
feat(microservices): ClientsModule.registerAsync

### PR DESCRIPTION
**Pull request by Amorites**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/nestjs/nest/blob/master/CONTRIBUTING.md
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] (TODO)Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Other... Please describe:
```

## What is the current behavior?

Issue Number: #4040

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

## Other information

### CODE REVIEW

File: packages/microservices/module/clients.module.ts [hunk 1]
```diff
@@ -15,4 +20,71 @@ export class ClientsModule {
       exports: clients,
     };
   }
+
+  static registerAsync(options: ClientsModuleAsyncOptions): DynamicModule {
+    const clients: Provider[] = (options || []).reduce(
```
Comment (kamilmysliwiec) [inline packages/microservices/module/clients.module.ts:25]: Can we rename it to `providers` instead?

Comment (kamilmysliwiec) [inline packages/microservices/module/clients.module.ts:25]:
```suggestion
    const clients: Provider[] = options.reduce(
```
I don't think we need this fallback?

File: packages/microservices/module/clients.module.ts [hunk 2]
```diff
@@ -15,4 +20,71 @@ export class ClientsModule {
...
+      (clients, item) =>
+        clients
+          .concat(this.createAsyncProviders(item))
+          .concat(item.extraProviders || []),
+      [],
+    );
+    const imports = (options || []).reduce(
```
Comment (kamilmysliwiec) [inline packages/microservices/module/clients.module.ts:32]: same as above (useless fallback?)

File: packages/microservices/module/clients.module.ts [hunk 3]
```diff
@@ -15,4 +20,71 @@ export class ClientsModule {
...
+      (res, option) =>
```
Comment (kamilmysliwiec) [inline packages/microservices/module/clients.module.ts:33]: `res` is somewhat ambiguous here

File: packages/microservices/module/clients.module.ts [hunk 4]
```diff
@@ -15,4 +20,71 @@ export class ClientsModule {
...
+      provide: options.name,
+      useFactory: this.factoryWrapper(
+        (optionsFactory: ClientsModuleOptionsFactory) =>
+          optionsFactory.createClientOptions(),
+      ),
+      inject: [options.useExisting || options.useClass],
+    };
+  }
+
+  private static factoryWrapper(
```
Comment (kamilmysliwiec) [inline packages/microservices/module/clients.module.ts:82]:
```suggestion
  private static createFactoryWrapper(
```

File: packages/microservices/module/clients.module.ts [hunk 5]
```diff
@@ -15,4 +20,71 @@ export class ClientsModule {
...
+    useFactory: ClientsProviderAsyncOptions['useFactory'],
+  ) {
+    return async (...args) => {
```
Comment (kamilmysliwiec) [inline packages/microservices/module/clients.module.ts:85]: Can we strongly type `...args` here?

File: packages/microservices/module/clients.module.ts [hunk 6]
```diff
@@ -15,4 +20,71 @@ export class ClientsModule {
...
+      const factoryReturnOption = await useFactory(...args);
```
Comment (kamilmysliwiec) [inline packages/microservices/module/clients.module.ts:86]:
```suggestion
      const clientOptions = await useFactory(...args);
```

**Comment by kamilmysliwiec** [comment]
Thank you! It will be added in the next release

**Comment by Artimunor** [comment]
Is this already added?
Also is there an example somewhere how to use it?
Thanks.

**Comment by ashkan-pm** [comment]
@Artimunor 
I couldn't find an example either, but I managed to get it working like this:
```js
imports: [
  ClientsModule.registerAsync([
    {
      name: 'AUTH_USER_SERVICE',
      imports: [ConfigModule],
      useFactory: async (configService: ConfigService) => ({
        transport: Transport.GRPC,
        options: {
          url: configService.get('USER_GRPC_URL'),
          protoPath: [
            join(configService.get('PROTO_PATH'), 'common.proto'),
            join(configService.get('PROTO_PATH'), 'user-service/auth-user.proto')
          ],
          package: ['common', 'auth']
        }
      }),
      inject: [ConfigService]
    }
  ]),
  ConfigModule
]
```

**Pull request merged**